agent/agent_endpoint.go

```diff
@@ -1217,79 +1214,45 @@ func (s *HTTPServer) AgentToken(resp http.ResponseWriter, req *http.Request) (in
-		changed := s.agent.tokens.UpdateUserToken(args.Token, token_store.TokenSourceAPI)
-		if changed {
-			triggerAntiEntropySync = true
-		}
-
-	case "acl_agent_token", "agent":
-		changed := s.agent.tokens.UpdateAgentToken(args.Token, token_store.TokenSourceAPI)
-		if changed {
-			triggerAntiEntropySync = true
-		}
-
-	case "acl_agent_master_token", "agent_master":
-		s.agent.tokens.UpdateAgentMasterToken(args.Token, token_store.TokenSourceAPI)
-
-	case "acl_replication_token", "replication":
-		s.agent.tokens.UpdateReplicationToken(args.Token, token_store.TokenSourceAPI)
-
-	default:
-		resp.WriteHeader(http.StatusNotFound)
-		fmt.Fprintf(resp, "Token %q is unknown", target)
-		return nil, nil
-	}
-
-	if triggerAntiEntropySync {
-		s.agent.sync.SyncFull.Trigger()
-	}
 
-	if s.agent.config.ACLEnableTokenPersistence {
-		tokens := persistedTokens{}
-
-		if tok, source := s.agent.tokens.UserTokenAndSource(); tok != "" && source == token_store.TokenSourceAPI {
-			tokens.Default = tok
-		}
+	err = s.agent.tokens.WithPersistenceLock(func() error {
+		triggerAntiEntropySync := false
+		switch target {
+		case "acl_token", "default":
+			changed := s.agent.tokens.UpdateUserToken(args.Token, token_store.TokenSourceAPI)
+			if changed {
+				triggerAntiEntropySync = true
+			}
 
-		if tok, source := s.agent.tokens.AgentTokenAndSource(); tok != "" && source == token_store.TokenSourceAPI {
-			tokens.Agent = tok
-		}
+		case "acl_agent_token", "agent":
+			changed := s.agent.tokens.UpdateAgentToken(args.Token, token_store.TokenSourceAPI)
+			if changed {
+				triggerAntiEntropySync = true
+			}
 
-		if tok, source := s.agent.tokens.AgentMasterTokenAndSource(); tok != "" && source == token_store.TokenSourceAPI {
-			tokens.AgentMaster = tok
-		}
+		case "acl_agent_master_token", "agent_master":
+			s.agent.tokens.UpdateAgentMasterToken(args.Token, token_store.TokenSourceAPI)
 
-		if tok, source := s.agent.tokens.ReplicationTokenAndSource(); tok != "" && source == token_store.TokenSourceAPI {
-			tokens.Replication = tok
-		}
+		case "acl_replication_token", "replication":
+			s.agent.tokens.UpdateReplicationToken(args.Token, token_store.TokenSourceAPI)
 
-		data, err := json.Marshal(tokens)
-		if err != nil {
-			s.agent.logger.Warn("failed to persist tokens", "error", err)
-			return nil, fmt.Errorf("Failed to marshal tokens for persistence: %v", err)
+		default:
+			resp.WriteHeader(http.StatusNotFound)
+			fmt.Fprintf(resp, "Token %q is unknown", target)
+			// TODO: return an error here to prevent persistence from happening.
+			return nil
 		}
 
-		if err := file.WriteAtomicWithPerms(filepath.Join(s.agent.config.DataDir, tokensPath), data, 0700, 0600); err != nil {
-			s.agent.logger.Warn("failed to persist tokens", "error", err)
-			return nil, fmt.Errorf("Failed to persist tokens - %v", err)
+		// TODO: is it safe to move this out of WithPersistenceLock?
+		if triggerAntiEntropySync {
+			s.agent.sync.SyncFull.Trigger()
 		}
```

Comment:
Calling out this TODO. I suspect it is safe to do without the lock, but I'd like to confirm.

Reply:
The token for SyncChanges is fetched from node state while holding statelock. Should be fine.